config/core/resources/trigger.yaml

```diff
@@ -101,6 +101,46 @@ spec:
                     description: 'Attributes filters events by exact match on event context attributes. Each key in the map is compared with the equivalent key in the event context. An event passes the filter if all values are equal to the specified values.  Nested context attributes are not supported as keys. Only string values are supported. '
                     type: object
                     x-kubernetes-preserve-unknown-fields: true
+              filters:
+                description: 'Filters is the filter to apply against all events from the Broker. Only events that pass this filter will be sent to the Subscriber. If not specified, will default to allowing all events. '
+                type: object
+                properties:
+                  exact:
+                    description: 'Exact evaluates to true if the value of the matching CloudEvents attribute is matches exactly the String value specified (case sensitive). Exact must contain exactly one property, where the key is the name of the CloudEvents attribute to be matched, and its value is the String value to use in the comparison. The attribute name and value specified in the filter express cannot be be empty strings.'
+                    type: object
+                    x-kubernetes-preserve-unknown-fields: true
+                  prefix:
+                    description: 'Prefix evaluates to true if the value of the matching CloudEvents attribute starts with the String value specified (case sensitive). Prefix must contain exactly one property, where the key is the name of the CloudEvents attribute to be matched, and its value is the String value to use in the comparison. The attribute name and value specified in the filter express cannot be be empty strings.'
+                    type: object
+                    x-kubernetes-preserve-unknown-fields: true
+                  suffix:
+                    description: 'Suffix evaluates to true if the value of the matching CloudEvents attribute ends with the String value specified (case sensitive). | Suffix must contain exactly one property, where the key is the name of the CloudEvents attribute to be matched, and its value is the String value to use in the comparison. he attribute name and value specified in the filter express cannot be be empty strings.'
+                    type: object
+                    x-kubernetes-preserve-unknown-fields: true
+                  not:
+                    description: 'Not evaluates to true if the nested expression evaluates to false.'
+                    type: object
+                    # Because kube doesn't allow to use $ref, we can't recursively define this schema.
```

Comment:
Copy pasted from somewhere else but irrelevant here it seems.

Reply:
Actually I think that's not copy/paste, this PR is a rewrite of #5205 and I believe what Francesco meant is that `x-kubernetes-preserve-unknown-fields: true` is used since `not`, `all` and `any` items are recursively referring to the same `filters` object, and k8s doesn't [support $ref for CRD schemas](https://github.com/kubernetes/kubernetes/issues/62872) .

Reply:
Is anything preventing us from just declaring `properties` for each object? I understand we may have to define the exact same properties multiple times, and it would have been _nice_ to use `$ref` to avoid that. But since this is not an option, I'm willing to have a strongly validated schema at the cost of duplication. (We already duplicate things like `spec.subscriber` in other CRDs, after all)

Curious to hear if other people feel strongly about the correctness of OpenAPI schemas or if the majority is fine with `x-kubernetes-preserve-unknown-fields` (bypass validation from the API server).

Reply:
@antoineco I'm not talking about code duplication, I'm talking about the fact that this is a recursive case. I haven't tested that, but can this be done without recursion?

Reply:
> Curious to hear if other people feel strongly about the correctness of OpenAPI schemas or if the majority is fine with x-kubernetes-preserve-unknown-fields (bypass validation from the API server).

In Eventing now schema validation is a requirement when it's possible, one of the agreements for v1 readiness. So for any feature to graduate to a stable release, a proper schema validation is needed as long as it's possible.

Reply:
Oh I see, I didn't notice that some `SubscriptionsAPIFilter` items were themselves containing the `SubscriptionsAPIFilter` type.

That complicates things indeed.